src/api/v2/types.rs

```diff
@@ -193,44 +200,231 @@ impl Reply for Status {
+		}
+
+		let decoded = hex::decode(&s[2..]).map_err(de::Error::custom)?;
+		let decoded_len = decoded.len();
+		let bytes: [u8; config::COMMITMENT_SIZE] = decoded
+			.try_into()
+			.map_err(|_| de::Error::invalid_length(decoded_len, &"Expected vector of 48 bytes"))?;
+
+		Ok(Commitment(bytes))
+	}
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+struct Extension {
+	rows: u16,
+	cols: u16,
+	data_root: Option<H256>,
+	commitments: Vec<Commitment>,
+	app_lookup: CompactDataLookup,
+}
+
+impl TryFrom<avail_subxt::primitives::Header> for Header {
+	type Error = anyhow::Error;
+
+	fn try_from(header: avail_subxt::primitives::Header) -> anyhow::Result<Self> {
+		Ok(Header {
+			hash: Encode::using_encoded(&header, blake2_256).into(),
+			parent_hash: header.parent_hash,
+			number: header.number,
+			state_root: header.state_root,
+			extrinsics_root: header.extrinsics_root,
+			extension: header.extension.try_into()?,
+		})
+	}
+}
+
+impl TryFrom<HeaderExtension> for Extension {
+	type Error = anyhow::Error;
+
+	fn try_from(value: HeaderExtension) -> Result<Self, Self::Error> {
+		match value {
+			HeaderExtension::V1(v1) => {
+				let commitments = commitments::from_slice(&v1.commitment.commitment)?
+					.into_iter()
+					.map(Commitment)
+					.collect::<Vec<_>>();
+				Ok(Extension {
+					rows: v1.commitment.rows,
+					cols: v1.commitment.cols,
+					data_root: Some(v1.commitment.data_root),
+					commitments,
+					app_lookup: v1.app_lookup,
+				})
+			},
+
+			HeaderExtension::V2(v2) => {
+				let commitments = commitments::from_slice(&v2.commitment.commitment)?
+					.into_iter()
+					.map(Commitment)
+					.collect::<Vec<_>>();
+
+				Ok(Extension {
+					rows: v2.commitment.rows,
+					cols: v2.commitment.cols,
+					data_root: v2.commitment.data_root,
+					commitments,
+					app_lookup: v2.app_lookup,
+				})
+			},
+		}
+	}
+}
+
+#[derive(Serialize, Clone)]
+#[serde(tag = "type", content = "message", rename_all = "kebab-case")]
+pub enum PublishMessage {
+	HeaderVerified(HeaderMessage),
+}
+
+impl TryFrom<PublishMessage> for Message {
+	type Error = anyhow::Error;
+	fn try_from(value: PublishMessage) -> Result<Self, Self::Error> {
+		serde_json::to_string(&value)
+			.map(ws::Message::text)
+			.context("Cannot serialize publish message")
+	}
 }
 
 pub type Sender = UnboundedSender<Result<ws::Message, warp::Error>>;
 
-pub struct Client {
+pub struct WsClient {
 	pub subscription: Subscription,
 	pub sender: Option<Sender>,
 }
 
-impl Client {
+impl WsClient {
 	pub fn new(subscription: Subscription) -> Self {
-		Client {
+		WsClient {
 			subscription,
 			sender: None,
 		}
 	}
+
+	fn is_subscribed(&self, topic: &Topic) -> bool {
+		self.subscription.topics.contains(topic)
+	}
 }
 
-pub type Clients = Arc<RwLock<HashMap<String, Client>>>;
+#[derive(Clone)]
+pub struct WsClients(pub Arc<RwLock<HashMap<String, WsClient>>>);
+
+impl WsClients {
+	pub async fn set_sender(&self, subscription_id: &str, sender: Sender) -> bool {
```

Comment:
Should this func return a `Result` maybe?

Reply:
It kind of awkward since caller checks if subscription exists and than tries to set it. Lock is not preserved between those two calls. But to make it more clear, I'll change it to result and log error instead of info. I guess this could be done better 🤷